### PR TITLE
fix: modified sales taxes and charges template

### PR DESCRIPTION
**Pull request by vishdha**
Peek 2020-05-14 14-41

### CODE REVIEW

**Review by DeveloperMetal — DISMISSED** [review]

File: erpnext/accounts/doctype/sales_taxes_and_charges_template/sales_taxes_and_charges_template.js [hunk 1]
```diff
@@ -5,3 +5,26 @@ cur_frm.cscript.tax_table = "Sales Taxes and Charges";
...
+
+frappe.ui.form.on("Sales Taxes and Charges", "category", function(doc, cdt, cdn) {
+	var d = locals[cdt][cdn];
+
+	if (d.category != 'Total' && d.add_deduct_tax == 'Deduct') {
+		frappe.msgprint(__("Cannot deduct when category is for 'Valuation' or 'Vaulation and Total'"));
+		d.add_deduct_tax = '';
+	}
+	refresh_field('add_deduct_tax', d.name, 'taxes');
+});
```
Comment (Alchez) [inline erpnext/accounts/doctype/sales_taxes_and_charges_template/sales_taxes_and_charges_template.js]:
```suggestion
frappe.ui.form.on("Sales Taxes and Charges", {
	add_deduct_tax: function (frm, cdt, cdn) {
		let row = locals[cdt][cdn];

		if (!row.category && row.add_deduct_tax) {
			frappe.msgprint(__("Please select Category first"));
			row.add_deduct_tax = '';
		} else if (row.category != 'Total' && row.add_deduct_tax == 'Deduct') {
			frappe.msgprint(__("Cannot deduct when category is not 'Total'"));
			row.add_deduct_tax = '';
		}

		refresh_field('add_deduct_tax', row.name, 'taxes');
	},

	category: function (frm, cdt, cdn) {
		let row = locals[cdt][cdn];

		if (row.category != 'Total' && row.add_deduct_tax == 'Deduct') {
			frappe.msgprint(__("Cannot deduct when category is not 'Total'"));
			row.add_deduct_tax = '';
		}

		refresh_field('add_deduct_tax', row.name, 'taxes');
	},
});
```

**Comment by Alchez** [comment]
@vishdha, can you move the category and add_deduct_tax logic inside the `accounts.js` file, below the other tax function:

https://github.com/DigiThinkIT/erpnext/blob/104cb3f3b8fbfbcc4bd27209be38848fb96ae291/erpnext/public/js/controllers/accounts.js#L12-L14

See how the `set_conditional_mandatory_rate_or_amount` function is setup

**Comment by Alchez** [comment]
Added upstream to [hotfix](https://github.com/frappe/erpnext/pull/21637) and [develop](https://github.com/frappe/erpnext/pull/21636) PRs.

**Comment by Alchez** [comment]
@neilLasrado, @vishdha, quick question: is Valuation even a thing for sales transacations? Should we just keep "Total" as the category option?

File: erpnext/public/js/controllers/accounts.js [hunk 2]
```diff
@@ -11,6 +11,7 @@ frappe.ui.form.on(cur_frm.doctype, {
 		$(frm.wrapper).on('grid-row-render', function(e, grid_row) {
 			if(in_list(['Sales Taxes and Charges', 'Purchase Taxes and Charges'], grid_row.doc.doctype)) {
 				erpnext.taxes.set_conditional_mandatory_rate_or_amount(grid_row);
+				erpnext.taxes.update_category_and_add_deduct_tax(grid_row);
```
Comment (Alchez) [inline erpnext/public/js/controllers/accounts.js]: Don't call this on setup, it's throwing an error on the console trying to refresh the table

File: erpnext/public/js/controllers/accounts.js [hunk 3]
```diff
@@ -246,6 +248,24 @@ if(!erpnext.taxes.flags[cur_frm.cscript.tax_table]) {
 		cur_frm.cscript.validate_taxes_and_charges(cdt, cdn);
 	});
 
+	frappe.ui.form.on(cur_frm.cscript.tax_table, "category", function(frm, cdt, cdn) {
+		if(open_form) {
```
Comment (Alchez) [inline erpnext/public/js/controllers/accounts.js:253]: Running the trigger causes the following issue: `ReferenceError: open_form is not defined`

**Review by Alchez — CHANGES_REQUESTED** [review]

File: erpnext/public/js/controllers/accounts.js [hunk 4]
```diff
@@ -246,6 +247,30 @@ if(!erpnext.taxes.flags[cur_frm.cscript.tax_table]) {
...
+	frappe.ui.form.on(cur_frm.cscript.tax_table, "category", function(frm, cdt, cdn) {
+		var d = locals[cdt][cdn];
+		var open_form = frm.open_grid_row();
+		if(open_form) {
+			erpnext.taxes.update_category_and_add_deduct_tax(open_form);
+		} else {
+			// apply in current row
+			erpnext.taxes.update_category_and_add_deduct_tax(frm.get_field('taxes').grid.get_row(cdn));
+		}
+		refresh_field('add_deduct_tax', d.name, 'taxes');
```
Comment (Alchez) [inline erpnext/public/js/controllers/accounts.js]: You already have the `open_form` object, no need for the `d` object, you can just do:

```suggestion
		refresh_field('add_deduct_tax', open_form.doc.name, 'taxes');
```

**Review by Alchez — APPROVED** [review]

**Pull request merged**